### ISSUE TITLE
[CS2] Destructuring invalid LHS generates invalid JS

### ISSUE THREAD

**Issue by connec** [jashkenas/coffeescript#4673]
See http://coffeescript.org/v2/#try:%7B%20%7Bb...%7D...%20%7D%20%3D%20c

```coffee
{ {b...}... } = c
````

Compiles to:

```js
({} = c);
Object.assign({}, b) = objectWithoutKeys(c, []);
```

I suspect we're missing something when verifying that an object is a valid LHS. Specifically, when a splat is present it's name must be a valid LHS.

**Comment by GeoffreyBooth** [comment]
@zdenko?

**Comment by connec** [comment]
I ended up fixing this whilst playing with an alternative fix for #4651 - https://github.com/jashkenas/coffeescript/compare/2...connec:fix-4651-alt

It will need rebased once #4652 is resolved.

**Comment by GeoffreyBooth** [comment]
@connec It’s actually a little worse than your initial comment. Look at the `var` declarations line:

```js
var Object.assign({}, b), b,
  objectWithoutKeys = function(o, ks) { var res = {}; for (var k in o) ([].indexOf.call(ks, k) < 0 && {}.hasOwnProperty.call(o, k)) && (res[k] = o[k]); return res; };

({} = c);
Object.assign({}, b) = objectWithoutKeys(c, []);
```

`var Object.assign({}, b), b`?

**Comment by helixbass** [comment]
Should be fixed by #4683. See its description, I think @connec's example here actually should compile, as eg `({...{...b}} = c)` seems to be valid ES6 (works in Chrome console)

**Comment by GeoffreyBooth** [comment]
Fixed by #4683.
